contractsV5/staking/StakingContract.sol

```diff
@@ -0,0 +1,286 @@
+        // compute epoch number
+        uint N = getCurrentEpochNumber();
+        address S = msg.sender;
+
+        require(latestStakeBalance[S] >= amount, "withdraw: latest amount staked is less than withdrawal amount");
+
+        initDataIfNeeded(S);
+
+        // actually not necessary, by right at here stakes[N+1][S] should be equal latestStake[S]
+        require(stakes[N + 1][S] >= amount, "withdraw: amount staked at next epoch is less than withdrawal amount");
+
+        stakes[N + 1][S] -= amount;
+        latestStakeBalance[S] -= amount;
+
+        address dAddr = delegatedAddress[N][S];
+
+        uint curStakes = stakes[N][S];
+        uint latestStakeBal = latestStakeBalance[S];
+        uint newStakes = curStakes <= latestStakeBal ? curStakes : latestStakeBal;
+        uint penaltyAmount = curStakes - newStakes; // newStakes is always <= curStakes
+
+        if (penaltyAmount > 0) {
+            if (dAddr != S) {
+                initDataIfNeeded(dAddr);
+                // S has delegated to dAddr, withdraw will affect his stakes + dAddr's delegated stakes
+                delegatedStakes[N][dAddr] -= penaltyAmount;
+            }
+            stakes[N][S] = newStakes;
+
+            if (address(DAO) != address(0)) {
+                // note: DAO needs to know which address to penalise
+                // if staker has delegated, address to penalise is delegated address
+                DAO.handleWithdrawal(dAddr, penaltyAmount);
+            }
+        }
+
+        dAddr = delegatedAddress[N + 1][S];
+        if (dAddr != S) {
+            initDataIfNeeded(dAddr);
+
+            require(delegatedStakes[N + 1][dAddr] >= amount, "withdraw: delegated stake is smaller than next epoch stake");
+            require(latestDelegatedStakes[dAddr] >= amount, "withdraw: latest delegated stake is smaller than next epoch stake");
+
+            delegatedStakes[N + 1][dAddr] -= amount;
+            latestDelegatedStakes[dAddr] -= amount;
+        }
+
+        // transfer KNC back to user
+        require(KNC_TOKEN.transfer(S, amount), "withdraw: can not transfer knc to the sender");
+
+        emit Withdrew(N, S, amount);
+    }
+
+    // init data if needed, then true staker's data for current epoch
+    // for safe, only allow calling this func from DAO address
+    // Note: should only call when staker voted
+    function initAndReturnStakerDataForCurrentEpoch(address staker)
+        public
+        returns(uint _stake, uint _delegatedStake, address _delegatedAddress)
+    {
+        require(msg.sender == address(DAO), "getStakerDataForCurrentEpoch: sender is not DAO address");
+
+        initDataIfNeeded(staker);
+
+        uint curEpoch = getCurrentEpochNumber();
+        _stake = stakes[curEpoch][staker];
+        _delegatedStake = delegatedStakes[curEpoch][staker];
+        _delegatedAddress = delegatedAddress[curEpoch][staker];
+    }
+
+    function getStakerDataForPastEpoch(address staker, uint epoch)
```

Comment:
Why do we have this function?

Reply:
When user wants to get his reward percentage or claim reward from DAO, we need to get staker's data (stake, delegated stake, delegated address) from Staking.
It's noted that if the data has not been inited, it means user didn't vote during this epoch

Reply:
I see, then returning 0 for everything seems ok